Move Scrollbar state reads out of composition

bentrengrove opened the pull request:
The way scrollbar was implemented, it observes all changes in composition. This is bad for scroll performance causing recomposition on every frame of scroll that doesn't need to happen.

This PR moves the work of observing the scroll state into the layout phase.

### Benchmarks Results
Iterations: 10 (so probably liable to a lot of noise)
#### Before
```
ScrollForYouFeedBenchmark_scrollFeedCompilationBaselineProfile
frameDurationCpuMs   P50   7.0,   P90   9.0,   P95  14.0,   P99  15.2
frameOverrunMs       P50  -5.7,   P90  -1.1,   P95   5.0,   P99   5.8
Traces: Iteration 0 1 2 3 4 5 6 7 8 9
```
#### After
```
ScrollForYouFeedBenchmark_scrollFeedCompilationBaselineProfile
frameDurationCpuMs   P50   3.7 (+89%),   P90  6.7 (+34%),   P95  10.7  (+30%),   P99  16.0 (-5%)
frameOverrunMs       P50  -8.3,   P90 -5.4,   P95  -1.0,   P99   5.5
Traces: Iteration 0 1 2 3 4 5 6 7 8 9
```

bentrengrove commented:
Ran out of time for the day, but keen for your thoughts, TJ.
I probably still need to clean this up sorry

tunjid commented on core/designsystem/src/main/kotlin/com/google/samples/apps/nowinandroid/core/designsystem/component/scrollbar/Scrollbar.kt:
```diff
@@ -320,84 +304,123 @@ fun Scrollbar(
+                        b = track.size - thumbSizePx,
+                    )
+
+                    val scrollbarThumbMovedDp = max(
```
The unit here is still pixels instead of dp, so maybe the val name should be `scrollbarThumbMovedPx`?

tunjid commented on core/designsystem/src/main/kotlin/com/google/samples/apps/nowinandroid/core/designsystem/component/scrollbar/Scrollbar.kt:
```diff
@@ -320,84 +304,123 @@ fun Scrollbar(
+
+                    val placeable = measurable.measure(constrained)
+                    layout(placeable.width, placeable.height) {
+                        placeable.place(offset)
```
Noice

tunjid commented:
LGTM, thanks so much for this! I was wondering, `ScrollbarState` itself is a value class for a packed `Float` primitive. For something that changes so frequently on scroll, I was trying to find a way to use `mutableFloatStateOf` for it's backing value instead of just `mutableStateOf`. Is there a way to do this?

bentrengrove commented:
> LGTM, thanks so much for this! I was wondering, `ScrollbarState` itself is a value class for a packed `Float` primitive. For something that changes so frequently on scroll, I was trying to find a way to use `mutableFloatStateOf` for it's backing value instead of just `mutableStateOf`. Is there a way to do this?

There sure is, I have updated the implementation with this. I'd love to know how much performance the packing you are doing is saving, I think its probably pretty minimal but I'm curious. You could probably save a bit of complexity just by having two 1mutableFloatStateOf` instead of packing into one `mutableLongState`.

I also added a custom Modifier.Node implementation that more efficiently updates it's color by not reading the color state in composition. You could actually go really far with this and move all the interaction observation into the Modifier.Node and then it would fully scroll with no recomposition needed. That was too much for today though.

I'm measuring the benchmark now to update the PR
